File: scripts/lib/tokens.js
/**
 * lib/tokens.js — Cargador y resolutor de la fuente de verdad (Node).
 *
 * Lee 03_tokens/tokens.json, resuelve las referencias entre llaves
 * (p. ej. "{color.naranja}") hasta su valor primitivo y expone los valores de
 * marca que los generadores NECESITAN, para que NINGÚN generador tenga colores
 * ni familias tipográficas quemados. Si un valor de marca cambia, se edita
 * tokens.json y todos los generadores lo heredan.
 *
 *   const { C, FONT, SZ, PT, twips, raw } = require("../lib/tokens.js");
 *
 * - C: colores en formato docx/openpyxl (HEX sin '#'), con las mismas claves
 *      que usaban los generadores del skill (primary, arena, green, …).
 * - FONT: familias { heading, body } (primer nombre de cada fontFamily token).
 * - SZ: tamaños tipográficos en MEDIOS PUNTOS (pt × 2), como los espera docx.
 * - PT: tamaños tipográficos en puntos.
 * - twips(pt): convierte puntos a veinteavos (docx spacing).
 * - raw(path): valor resuelto de cualquier token por su ruta.
 */
const fs = require("fs");
const path = require("path");

const TOKENS_PATH = path.join(__dirname, "..", "..", "03_tokens", "tokens.json");
const tokens = JSON.parse(fs.readFileSync(TOKENS_PATH, "utf8"));

// --- aplanar a { "color.naranja": "#E94513", ... } ---
const leaves = {};
(function walk(node, prefix) {
  if (node && typeof node === "object") {
    if ("$value" in node) { leaves[prefix.join(".")] = node.$value; return; }
    for (const k of Object.keys(node)) {
      if (k.startsWith("$")) continue;
      walk(node[k], prefix.concat(k));
    }
  }
})(tokens, []);

// --- resolver referencias {a.b.c} de forma recursiva ---
const REF = /^\{([^}]+)\}$/;
function resolve(value, seen = []) {
  if (typeof value === "string") {
    const m = value.trim().match(REF);
    if (m) {
      const target = m[1];
      if (seen.includes(target)) throw new Error("ciclo de referencia: " + seen.concat(target).join(" -> "));
      if (!(target in leaves)) throw new Error("referencia rota: {" + target + "}");
      return resolve(leaves[target], seen.concat(target));
    }
  }
  return value;
}
const raw = (p) => resolve(leaves[p]);
const hex = (p) => String(raw(p)).replace(/^#/, ""); // docx/openpyxl no usan '#'

// --- C: mismas claves que los generadores del skill, ahora desde tokens ---
const C = {
  primary:     hex("color.naranja"),
  arena:       hex("color.arena"),
  green:       hex("color.verde"),
  black:       hex("color.negro"),
  white:       hex("color.blanco"),
  grayText:    hex("color.gris-texto"),
  grayMedium:  hex("color.gris-borde"),
  grayLight:   hex("color.gris-linea"),
  lightOrange: hex("color.durazno"),
  lightSand:   hex("color.arena-claro"),
  lightGreen:  hex("color.verde-claro"),
};

const FONT = {
  heading: raw("font.family.heading")[0], // "Space Grotesk"
  body:    raw("font.family.body")[0],    // "Calibri"
};

// tamaños en puntos desde font.size.* (objeto {value, unit})
const ptOf = (p) => raw(p).value;
const PT = {
  coverTitle: ptOf("font.size.coverTitle"), h1: ptOf("font.size.h1"), h2: ptOf("font.size.h2"),
  h3: ptOf("font.size.h3"), body: ptOf("font.size.body"), caption: ptOf("font.size.caption"),
  footer: ptOf("font.size.footer"),
};
const SZ = Object.fromEntries(Object.entries(PT).map(([k, v]) => [k, v * 2])); // docx: medios puntos
const twips = (pt) => Math.round(pt * 20);

module.exports = { C, FONT, SZ, PT, twips, raw, hex, tokensPath: TOKENS_PATH };

File: scripts/lib/tokens.py
"""
lib/tokens.py — Cargador y resolutor de la fuente de verdad (Python).

Espejo de lib/tokens.js para los generadores Python (xlsx) y el verificador
check_fucai.py. Lee 03_tokens/tokens.json, resuelve las referencias {a.b.c} y
expone los valores de marca para que NINGÚN script Python tenga colores ni
familias quemados.

    import sys, os
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "lib"))
    from tokens import C, FONT, PT, raw

- C: colores en HEX sin '#' (openpyxl), mismas claves que el skill.
- FONT: {"heading": "Space Grotesk", "body": "Calibri"}.
- PT: tamaños tipográficos en puntos (escala Word/PDF).
- raw(path): valor resuelto de cualquier token.
"""
import json, os, re

_TOKENS_PATH = os.path.normpath(os.path.join(os.path.dirname(__file__), "..", "..", "03_tokens", "tokens.json"))
with open(_TOKENS_PATH, encoding="utf-8") as _f:
    _tokens = json.load(_f)

_leaves = {}
def _walk(node, prefix):
    if isinstance(node, dict):
        if "$value" in node:
            _leaves[".".join(prefix)] = node["$value"]; return
        for k, v in node.items():
            if k.startswith("$"):
                continue
            _walk(v, prefix + [k])
_walk(_tokens, [])

_REF = re.compile(r"^\{([^}]+)\}$")
def _resolve(value, seen=None):
    seen = seen or []
    if isinstance(value, str):
        m = _REF.match(value.strip())
        if m:
            target = m.group(1)
            if target in seen:
                raise ValueError("ciclo de referencia: " + " -> ".join(seen + [target]))
            if target not in _leaves:
                raise KeyError("referencia rota: {%s}" % target)
            return _resolve(_leaves[target], seen + [target])
    return value

def raw(path):
    return _resolve(_leaves[path])

def _hex(path):
    return str(raw(path)).lstrip("#")

C = {
    "primary":     _hex("color.naranja"),
    "arena":       _hex("color.arena"),
    "green":       _hex("color.verde"),
    "black":       _hex("color.negro"),
    "white":       _hex("color.blanco"),
    "grayText":    _hex("color.gris-texto"),
    "grayMedium":  _hex("color.gris-borde"),
    "grayLight":   _hex("color.gris-linea"),
    "lightOrange": _hex("color.durazno"),
    "lightSand":   _hex("color.arena-claro"),
    "lightGreen":  _hex("color.verde-claro"),
}

FONT = {"heading": raw("font.family.heading")[0], "body": raw("font.family.body")[0]}

PT = {k: raw("font.size.%s" % k)["value"] for k in
      ("coverTitle", "h1", "h2", "h3", "body", "caption", "footer")}

TOKENS_PATH = _TOKENS_PATH
